light/detector.go

```diff
@@ -0,0 +1,223 @@
+// it has received from another and preforms verifySkipping at the heights of each of the intermediate
+// headers in the trace until it reaches the divergentHeader. 1 of 2 things can happen.
+//
+// 1. The light client verifies a header that is different to the intermediate header in the trace. This
+//    is the bifurcation point and the light client can create evidence from it
+// 2. The source stops responding, doesn't have the block or sends an invalid header in which case we
+//    return the error and remove the witness
+func (c *Client) examineConflictingHeaderAgainstTrace(
+	trace []*types.LightBlock,
+	divergentHeader *types.SignedHeader,
+	source provider.Provider, now time.Time) ([]*types.LightBlock, *types.LightBlock, error) {
+
+	var previouslyVerifiedBlock *types.LightBlock
+
+	for idx, traceBlock := range trace {
+		// The first block in the trace MUST be the same to the light block that the source produces
+		// else we cannot continue with verification.
+		sourceBlock, err := source.LightBlock(traceBlock.Height)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if idx == 0 {
+			if shash, thash := sourceBlock.Hash(), traceBlock.Hash(); !bytes.Equal(shash, thash) {
+				return nil, nil, fmt.Errorf("trusted block is different to the source's first block (%X = %X)",
+					thash, shash)
+			}
+			previouslyVerifiedBlock = sourceBlock
+			continue
+		}
+
+		// we check that the source provider can verify a block at the same height of the
+		// intermediate height
+		trace, err := c.verifySkipping(source, previouslyVerifiedBlock, sourceBlock, now)
+		if err != nil {
+			return nil, nil, fmt.Errorf("verifySkipping of conflicting header failed: %w", err)
+		}
+		// check if the headers verified by the source has diverged from the trace
+		if shash, thash := sourceBlock.Hash(), traceBlock.Hash(); !bytes.Equal(shash, thash) {
+			// Bifurcation point found!
+			return trace, traceBlock, nil
+		}
+
+		// headers are still the same. update the previouslyVerifiedBlock
+		previouslyVerifiedBlock = sourceBlock
+	}
+
+	// We have reached the end of the trace without observing a divergence. The last header  is thus different
+	// from the divergent header that the source originally sent us, then we return an error.
+	return nil, nil, fmt.Errorf("source provided different header to the original and is the same as the targer"+
+		"header of the trace (%X != %X = %X)",
+		previouslyVerifiedBlock.Hash(), divergentHeader.Hash(), trace[len(trace)-1].Hash())
+
+}
+
+// IsInvalidHeader takes a trusted header and matches it againt a conflicting header
+// to determine whether the conflicting header was the product of a valid state transition
+// or not. If it is then all the deterministic fields of the header should be the same.
+// If not, it is an invalid header and constitutes a lunatic attack.
+func IsInvalidHeader(trusted, conflicting *types.Header) bool {
```

Comment:
do we really want to expose this as public API? the fact that it's public in spec does not mean we should make it public in Go API. spec != code